### ISSUE TITLE
Problems reading XLSX files generated by Crystal Reports

### ISSUE THREAD

**Issue by cmanley** [box/spout#260]
I'm using the latest master version and I'm getting this exception with an XLSX file exported by Crystal Reports:
Box\Spout\Common\Exception\IOException with message:
Could not open test.xlsx for reading! (Index invalid or out of range)
in Spout/Reader/AbstractReader.php line 105

It could be related to:
https://github.com/box/spout/issues/254

If I open the xlsx file in LibreOffice and save it again, then it works fine. 

When reading other Crystal Reports xlsx files, I had other weird things happen such as custom format strings being displayed instead of the actual values in some fields, and in some other fields empty strings were returned when there were really values in those fields. But this may show up if the reported exception problem is solved.
Is there a way to send the small file privately because it contains real production (possibly somewhat sensitive) data? I get the files by proxy, so that's why I can't easily get test data.

**Comment by adrilo** [comment]
Hi @cmanley, it seems like Crystal Reports is causing a lot of issues ! :)
The line you indicated is where the exception is thrown and the error message is pretty vague. Without a stacktrace or (even better) a test file, I won't be able to help you. If you could get me one or the other, it would be a good start.

**Comment by cranesoft** [comment]
I too had an issue reading a Crystal Reports saved file.  In my case it was an .xls file and the issue ended up being that Crystal saved the file in HTML slapping on an .xls extension.  Similar to you I could open and save the file in something else (Excel in my case) and box / spout had no issue.

**Comment by cranesoft** [comment]
Further, I determined the problem by opening up my file in an editor.  The file was clearly HTML, not binary.

**Comment by cmanley** [comment]
@adrilo I can mail you a test file of 10 rows but I don't know your mail address.
@cranesoft In my case it was really xlsx.

**Comment by cmanley** [comment]
There wasn't much of a useful stack trace due to the try-catch construct around line 105, so I commented out the try-catch parts without changing the line numbers. This is the error and stack trace that I get now:
`PHP Fatal error:  Uncaught exception 'RuntimeException' with message 'Index invalid or out of range' in ...Spout/Reader/XLSX/Helper/SharedStringsCaching/InMemoryStrategy.php:42
`
Stack trace:
`#0 ...Spout/Reader/XLSX/Helper/SharedStringsCaching/InMemoryStrategy.php(42): SplFixedArray->offsetSet(0, 'OP_VERZOPDRNR')
#1 ...Spout/Reader/XLSX/Helper/SharedStringsHelper.php(121): Box\Spout\Reader\XLSX\Helper\SharedStringsCaching\InMemoryStrategy->addStringForIndex('OP_VERZOPDRNR', 0)
#2 ...Spout/Reader/XLSX/Reader.php(69): Box\Spout\Reader\XLSX\Helper\SharedStringsHelper->extractSharedStrings()
#3 ...Spout/Reader/AbstractReader.php(102): Box\Spout\Reader\XLSX\Reader->openReader('test.xlsx in ...Spout/Reader/XLSX/Helper/SharedStringsCaching/InMemoryStrategy.php on line 42
`

**Comment by cmanley** [comment]
Attached is a small Crystal Reports XLSX file that causes the aforementioned exception.
[CrystalReports.xlsx.zip](https://github.com/box/spout/files/347832/CrystalReports.xlsx.zip)

**Comment by adrilo** [comment]
@cmanley, I can't even open the file in Excel (Mac version). It seems like there is something wrong with the generation of the file. Are you able to open it with your version of Excel?

**Comment by cmanley** [comment]
Yes I can open it using Libre Office. I assume you renamed the zip extension to xlsx.
-------- Original message --------From: Adrien Loison <notifications@github.com> Date: 05/07/2016  18:15  (GMT+01:00) To: box/spout <spout@noreply.github.com> Cc: Craig Manley <cmanley@xs4all.nl>, Mention <mention@noreply.github.com> Subject: Re: [box/spout] Problems reading XLSX files generated by Crystal
  Reports (#260) 
@cmanley, I can't even open the file in Excel (Mac version). It seems like there is something wrong with the generation of the file. Are you able to open it with your version of Excel?

**Comment by adrilo** [comment]
Indeed it works with LibreOffice. After investigating, the issue comes from the fact that in the XML storing the shared strings, the `uniqueCount` and `count` of shared strings is not specified.
In a previous issue (#254), I changed the way I get the count by looking at `count` when `uniqueCount` is missing. It seems like I need to add something else to take into account the absence of these properties.

I looked at the spec to be sure (as I don't want to add unneeded complexity) but it says that both params are optional (although when one is used, the other one is required).

I'll push a fix hopefully next week.
